### PR TITLE
BUG: .iloc[:] and .loc[:] return a copy of the original object #13873

### LINKED ISSUE #13873: .iloc[:] and .loc[:] return the original object
Each of these asserts should fail:

``` python
# on pandas 0.18.1
import pandas as pd

df = pd.DataFrame({'x': range(3)})
assert df.iloc[:] is df
assert df.loc[:] is df

series = df.x
assert series.iloc[:] is series
assert series.loc[:] is series
```

Instead, for the sake of predictability, these should always return a new DataFrame/Series, even if they reuse the same data. See [here](https://github.com/pydata/pandas/issues/13842#issuecomment-236685599) for an example of how this resulted in a user facing bug within pandas.

Thanks @mborysow for pointing this out.

### PR DESCRIPTION
**Pull request by margaret**
Previously they returned the original. This PR should make it consistent with numpy and list.

 - [x] closes #13873
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
pls run the entire asv suite. I think you are copying unecessarily in a few places. This needs to be a *very* narrow change.

IOW, *only* when a copy is specify requested on the entire DataFrame, e.g. ``df.iloc[:]`` or ``df.loc[:]``. NOT anyother time as they return view by-definition.

**Comment by margaret** [comment]
@jreback Okay, I can revert the change for the tuple cases. I was going off of this comment on the original issue https://github.com/pandas-dev/pandas/issues/13873#issuecomment-236717538

**Comment by jreback** [comment]
@margaret yes, this is *only* for all null slices passed. You have to be *very* careful with copying as that can really change performance. We always return views.

**Comment by chris-b1** [comment]
I think we should probably only be shallow copying in these cases (`.copy(deep=False)`)  - seems to be consistent with numpy semantics.

```
a = np.array([1, 2, 3])

a[:] is a
Out[44]: False

In [46]: a[:].base is a
Out[46]: True
```

**Comment by margaret** [comment]
Ran all the tests locally (but only on Python 2.7), currently failing two tests: [`test_sorting.py`](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_sorting.py#L84) for `Series` that was implemented for issue https://github.com/pandas-dev/pandas/issues/5856 which was fixed with PR https://github.com/pandas-dev/pandas/issues/5853

```
        # GH 5856/5853
        # Series.sort_values operating on a view
        df = DataFrame(np.random.randn(10, 4))
        s = df.iloc[:, 0]

        def f():
            s.sort_values(inplace=True)

        pytest.raises(ValueError, f)
```
and something in `check_round_trip_frame` in [`test_clipboard.py`](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_clipboard.py#L69)
```
E       AssertionError: DataFrame are different
E       
E       DataFrame shape mismatch
E       [left]:  (2, 2)
E       [right]: (0, 1)

pandas/util/testing.py:1109: AssertionError
```

Looking into these.

**Comment by chris-b1** [comment]
The clipboard one is a flaky test, I'd ignore it - other failure probably is legitimate.

Can you also run the performance benchmarks to see if anything was impacted? Docs here:
http://pandas.pydata.org/pandas-docs/stable/contributing.html#running-the-performance-test-suite

**Comment by margaret** [comment]
Yep, looks like it is just the sorting a view test that's failing.

**Comment by margaret** [comment]
Ah, it's because when calling `df.iloc[:,0]`, down in 
```
-> s = df.iloc[:, 0]
  /Users/margaret/Documents/Projects/pandas/pandas/core/indexing.py(1325)__getitem__()
-> return self._getitem_tuple(key)
  /Users/margaret/Documents/Projects/pandas/pandas/core/indexing.py(1664)_getitem_tuple()
-> return self._getitem_lowerdim(tup)
> /Users/margaret/Documents/Projects/pandas/pandas/core/indexing.py(992)_getitem_lowerdim()
-> return getattr(section, self.name)[new_key]
```
when parsing the `0`, it ends up calling and returning `iloc[:]`.

**Comment by margaret** [comment]
Python 3.6 tests passed locally

**Comment by margaret** [comment]
Travis failure is on `ci/lint.sh`, but `git diff upstream/master --name-only -- '*.py' | flake8 --diff` passes, going to try running that locally.

**Comment by margaret** [comment]
Also the benchmarks are about 2/3 done now, but there have been 3 failures. Going to look into that but if anyone has thoughts please share.
```
[  0.00%] · For pandas commit hash b904308f:
...
[ 33.22%] ··· Running panel_methods.PanelMethods.time_pct_change_items 
[ 33.29%] ··· Running panel_methods.PanelMethods.time_pct_change_major
[ 33.36%] ··· Running panel_methods.PanelMethods.time_pct_change_minor
```

**Comment by margaret** [comment]
asv benchmark results in this [gist](https://gist.github.com/margaret/489f8b5a8e6e8eb754a01df0c425dbb7).

File: pandas/tests/indexing/test_iloc.py [hunk 1]
```diff
@@ -591,3 +591,12 @@ def test_iloc_empty_list_indexer_is_ok(self):
         tm.assert_frame_equal(df.iloc[[]], df.iloc[:0, :],
                               check_index_type=True,
                               check_column_type=True)
+
+    def test_loc_identity_slice_returns_new_object(self):
+        # GH13873
+        df = DataFrame({'a': [1, 2, 3]})
+        result = df.iloc[:]
+        assert not result is df
+        # should be a shallow copy
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: can you add a blank line

File: pandas/tests/indexing/test_iloc.py [hunk 2]
```diff
@@ -591,3 +591,12 @@ def test_iloc_empty_list_indexer_is_ok(self):
         tm.assert_frame_equal(df.iloc[[]], df.iloc[:0, :],
                               check_index_type=True,
                               check_column_type=True)
+
+    def test_loc_identity_slice_returns_new_object(self):
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: just rename to ``test_identity_slice_returns_new_object`` (and same for one in the ``test_loc`` file

File: pandas/tests/indexing/test_loc.py [hunk 3]
```diff
@@ -630,3 +630,12 @@ def test_loc_empty_list_indexer_is_ok(self):
         tm.assert_frame_equal(df.loc[[]], df.iloc[:0, :],
                               check_index_type=True,
                               check_column_type=True)
+
+    def test_loc_identity_slice_returns_new_object(self):
+        # GH13873
+        df = DataFrame({'a': [1, 2, 3]})
+        result = df.loc[:]
+        assert not result is df
+        # should be a shallow copy
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: blank line

File: pandas/tests/indexing/test_loc.py [hunk 4]
```diff
@@ -630,3 +630,12 @@ def test_loc_empty_list_indexer_is_ok(self):
         tm.assert_frame_equal(df.loc[[]], df.iloc[:0, :],
                               check_index_type=True,
                               check_column_type=True)
+
+    def test_loc_identity_slice_returns_new_object(self):
+        # GH13873
+        df = DataFrame({'a': [1, 2, 3]})
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: can you test also for Series

File: pandas/tests/indexing/test_loc.py [hunk 5]
```diff
@@ -630,3 +630,12 @@ def test_loc_empty_list_indexer_is_ok(self):
...
+        result = df.loc[:]
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: also can you add a couple of tests (can be in the same function is ok) that assert that things are NOT a copy (
df.iloc[:, 0] for example (IOW want to test the counter-examples)

**Review by jreback — APPROVED** [review]
lgtm. want to cover the bases with a few more tests.

**Comment by TomAugspurger** [comment]
@margaret a couple (very tiny) linting errors :) https://travis-ci.org/pandas-dev/pandas/jobs/235411163

```
pandas/tests/indexing/test_iloc.py:599:16: E714 test for object identity should be 'is not'
pandas/tests/indexing/test_loc.py:638:16: E714 test for object identity should be 'is not'
```

File: pandas/tests/indexing/test_loc.py (hunk 5 above)
Comment (margaret) [inline pandas/tests/indexing/test_loc.py]: Double checking — slicing with `loc` should return the original object, but `iloc` should return a copy (that's the behavior in v0.20.1.)? If so, why does `iloc` return a copy?

**Comment by jreback** [comment]
both loc and iloc should return shallow copies for [:]

**Comment by margaret** [comment]
Indexing with a single loc or iloc call otherwise should always return a view, right? In v0.21.0 `df.iloc[:,0]` on a df created from a 2d ndarray returns a different object than `df[0]`. So is that actually incorrect behavior?

**Comment by jreback** [comment]
> Indexing with a single loc or iloc call otherwise should always return a view, right? In v0.21.0 df.iloc[:,0] on a df created from a 2d ndarray returns a different object than df[0]. So is that actually incorrect behavior?

can you show a specific example.

**Comment by margaret** [comment]
```
>>> pd.__version__
'0.20.1'
>>> df = pd.DataFrame(np.random.randn(10, 4))
>>> df.loc[:,0] is df[0]
True
>>> df.iloc[:,0] is df[0]
False
```

**Comment by jreback** [comment]
@margaret hmm, those should both be views. was this changed from 0.19.2?

**Comment by margaret** [comment]
Nope, looks like it's the same in 0.19.2.

**Comment by jreback** [comment]
yeah we prob don't systematically check this
welcome for u to open a new issue / PR for that

**Comment by jreback** [comment]
@margaret lgtm. can you add the same tests for ``df[:]`` and ``s[:]`` they should do the same.

**Comment by margaret** [comment]
CircleCI failed with
```
Suppressing install of ssh keys for fork PR builds
fatal: No such section!
fatal: No such section!
```
on 'start container' and then when it tried ssh-ing into the VM.  Could someone try re-running it? Or maybe there's a setting that needs to be updated?

**Comment by margaret** [comment]
@jreback

**Comment by jreback** [comment]
looks reasonable. pls rebase and push again to have the CI run fully.

**Comment by jreback** [comment]
ping on green.

**Comment by margaret** [comment]
@jreback ✅

**Comment by jreback** [comment]
thanks @margaret nice fix. sometimes it takes a while to get these in. keep em coming!

**Pull request merged**